pkg/piecestore/rpc/server/psdb/psdb.go

```diff
@@ -0,0 +1,246 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package psdb
+
+import (
+	"database/sql"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+
+	_ "github.com/mattn/go-sqlite3" // sqlite is weird and needs underscore
+
+	"storj.io/storj/pkg/piecestore"
+	pb "storj.io/storj/protos/piecestore"
+)
+
+// PSDB -- Piecestore database
+type PSDB struct {
+	DB  *sql.DB
+	mtx sync.Mutex
+}
+
+// OpenPSDB -- opens PSDB at DBPath
+func OpenPSDB(DBPath string) (*PSDB, error) {
+	if err := os.MkdirAll(filepath.Dir(DBPath), 0700); err != nil {
+		return nil, err
+	}
+
+	db, err := sql.Open("sqlite3", fmt.Sprintf("file:%s?cache=shared&mutex=full", DBPath))
+	if err != nil {
+		return nil, err
+	}
+
+	tx, err := db.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = tx.Exec("CREATE TABLE IF NOT EXISTS `ttl` (`id` TEXT UNIQUE, `created` INT(10), `expires` INT(10));")
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = tx.Exec("CREATE TABLE IF NOT EXISTS `bandwidth_agreements` (`agreement` BLOB, `signature` BLOB);")
+	if err != nil {
+		return nil, err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return nil, err
+	}
+
+	return &PSDB{DB: db}, nil
+}
+
+// deleteEntries -- checks for and deletes expired TTL entries
+func deleteEntriesFromFS(dir string, rows *sql.Rows) error {
+
+	for rows.Next() {
+		var expID string
+
+		err := rows.Scan(&expID)
+		if err != nil {
+			return err
+		}
+
+		// delete file on local machine
+		err = pstore.Delete(expID, dir)
+		if err != nil {
+			return err
+		}
+
+		log.Printf("Deleted file: %s\n", expID)
+		if rows.Err() != nil {
+			return rows.Err()
+		}
+	}
+
+	return nil
+}
+
+func (psdb *PSDB) Close() error {
+	return psdb.DB.Close()
+}
+
+// CheckEntries -- go routine to check ttl database for expired entries
+// pass in database and location of file for deletion
+func (psdb *PSDB) CheckEntries(dir string) error {
+
+	tickChan := time.NewTicker(time.Second * 5).C
+	for {
+		select {
+		case <-tickChan:
+			now := time.Now().Unix()
+
+			rows, err := psdb.DB.Query(fmt.Sprintf("SELECT id FROM ttl WHERE expires < %d AND expires > 0", now))
+			if err != nil {
+				return err
+			}
+			defer func() {
```

Comment:
defers only run when the function exits. so every loop through the for loop we add a defer without it ever running. huge stack of defers real quick